source/common/compressor/compressor_impl.cc

```diff
@@ -0,0 +1,108 @@
+#include "common/compressor/compressor_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Compressor {
+
+CompressorImpl::CompressorImpl()
+    : zstream_ptr_(new z_stream(), [](z_stream* z) {
+        deflateEnd(z);
+        delete z;
+      }) {
+  zstream_ptr_->zalloc = Z_NULL;
+  zstream_ptr_->zfree = Z_NULL;
+  zstream_ptr_->opaque = Z_NULL;
+}
+
+CompressorImpl::~CompressorImpl() {}
+
+void CompressorImpl::init(CompressionLevel comp_level, CompressionStrategy comp_strategy,
+                          int8_t window_bits, uint8_t memory_level = 8) {
+  if (initialized_) {
+    throw EnvoyException("Compressor can only be initialized once");
+  }
+  const int result = deflateInit2(zstream_ptr_.get(), static_cast<int>(comp_level), Z_DEFLATED,
+                                  static_cast<int>(window_bits), static_cast<int>(memory_level),
+                                  static_cast<int>(comp_strategy));
+  if (result < 0) {
+    throw EnvoyException(
```

Comment:
IMO it's not worth using exceptions here. I would just RELEASE_ASSERT. This should never fail.